setup/classes/class.ilCtrlStructureReader.php

```diff
@@ -102,146 +102,189 @@ function readStructure($a_force = false, $a_dir = "", $a_comp_prefix = "",
 	function read($a_cdir)
 	{
 		$ilDB = $this->getDB();
-		$il_absolute_path = realpath(dirname(__FILE__) .'/../../');
+		$il_absolute_path = realpath(dirname(__FILE__) . '/../../');
 
-		// check wether $a_cdir is a directory
-		if (!@is_dir($a_cdir))
-		{
+		$a_cdir = preg_replace('#//#', '/', $a_cdir);
+		if (!is_dir($a_cdir) || !is_readable($a_cdir)) {
 			return false;
 		}
 
-		// read current directory
-		$dir = opendir($a_cdir);
+		$directory_iter = new RecursiveDirectoryIterator(
+			$a_cdir,
+			FilesystemIterator::SKIP_DOTS
+		);
 
-		while($file = readdir($dir))
+		$filtered_directory_iter = new class($directory_iter, $il_absolute_path) extends RecursiveFilterIterator
```

Comment:
I know that the current code is quite old school. But anonymous classes are suspicious to me. Complex to read the code, and up to now I saw little use cases for these (maybe mocks in tests...).